waspc/src/Analyzer/TypeChecker/Internal.hs

```diff
@@ -0,0 +1,232 @@
+-- created would be @[("Todo", DeclType "app")]@.
+--
+-- = Statement Type Checking
+--
+-- In this second phase, the types of the argument to each declaration are checked
+-- to ensure they are valid for the declaration type. The implementation of the
+-- type inference rules is in "checkExpr", "unifyTypes", and "weaken".
+module Analyzer.TypeChecker.Internal
+  ( check,
+    hoistDeclarations,
+    checkAST,
+    checkStmt,
+    checkExpr,
+    unify,
+    unifyTypes,
+    weaken,
+  )
+where
+
+import Analyzer.Parser (AST)
+import qualified Analyzer.Parser as P
+import Analyzer.Type
+import Analyzer.TypeChecker.AST
+import Analyzer.TypeChecker.Monad
+import Analyzer.TypeChecker.TypeError
+import qualified Analyzer.TypeDefinitions as TD
+import Control.Arrow (left)
+import Control.Monad (foldM)
+import Data.Foldable (foldl')
+import qualified Data.HashMap.Strict as M
+import Data.List.NonEmpty (NonEmpty ((:|)), nonEmpty, toList)
+
+check :: AST -> T TypedAST
+check ast = hoistDeclarations ast >> checkAST ast
+
+hoistDeclarations :: AST -> T ()
+hoistDeclarations (P.AST stmts) = mapM_ hoistDeclaration stmts
+  where
+    hoistDeclaration :: P.Stmt -> T ()
+    hoistDeclaration (P.Decl typName ident _) = setType ident $ DeclType typName
+
+checkAST :: AST -> T TypedAST
+checkAST (P.AST stmts) = TypedAST <$> mapM checkStmt stmts
+
+checkStmt :: P.Stmt -> T TypedStmt
+checkStmt (P.Decl typName name expr) =
+  lookupDeclType typName >>= \case
+    Nothing -> throw $ NoDeclarationType typName
+    Just (TD.DeclType _ expectedType) -> do
+      -- Decides whether the argument to the declaration has the correct type
+      mTypedExpr <- weaken expectedType <$> checkExpr expr
+      case mTypedExpr of
+        Left e -> throw e
+        Right typedExpr -> return $ Decl name typedExpr (DeclType typName)
+
+-- | Determine the type of an expression, following the inference rules described in
+-- the wasplang document. Some these rules are referenced by name in the comments
+-- of the following functions using [Brackets].
+checkExpr :: P.Expr -> T TypedExpr
+checkExpr (P.StringLiteral s) = return $ StringLiteral s
+checkExpr (P.IntegerLiteral i) = return $ IntegerLiteral i
+checkExpr (P.DoubleLiteral d) = return $ DoubleLiteral d
+checkExpr (P.BoolLiteral b) = return $ BoolLiteral b
+checkExpr (P.ExtImport n s) = return $ ExtImport n s
+checkExpr (P.Identifier ident) =
+  lookupType ident >>= \case
+    Nothing -> throw $ UndefinedIdentifier ident
+    Just typ -> return $ Var ident typ
+-- For now, the two quoter types are hardcoded here, it is an error to use a different one
+-- TODO: this will change when quoters are added to "Analyzer.TypeDefinitions"
+checkExpr (P.Quoter "json" s) = return $ JSON s
+checkExpr (P.Quoter "psl" s) = return $ PSL s
+checkExpr (P.Quoter tag _) = throw $ QuoterUnknownTag tag
+-- The type of a list is the unified type of its values.
+-- This poses a problem for empty lists, there is not enough information to choose a type.
```

Comment:
What are the consequences of implementing this, pros and cons? If nothing major, I am in for it. Also, pls provide rich comments with explanation of different choices, why we made this one, and any considerations for the future.

Reply:
Ok sure, I have to admit I am not able to judge this completely because I haven't given it enough thought, but I trust your judgement, so let's go for it, let's just make sure to document it well.
You didn't answer my question: what was the more complicated alternative that you were thinking about? If that alternative might be interesting for the future, let's also document that.

Reply:
Ok, let's do it like that!